### PR TITLE
Build Python Wheel from pyoorb

### PR DESCRIPTION
**Pull request by spenczar**
This is a first crack at the basics for #135. It adds a `python/setup.py` file which hooks into setuptools to allow building a binary distributable wheel.

It relies on the existing Makefile to do all the heavy lifting; it just copies the resulting `pyoorb.[...].so` and packages it as a wheel.

Some obvious todos to accomplish #135:
 - The `setup()` call should be filled out with more info: a project description, list of maintainers, and so on.
 - This should be tested on OSX and other Python versions - I have only done it on Linux and with Python 3.8
 - Need to also add infrastructure to build it in github actions, and upload releases to PyPI.
 - I am a little bit worried about the wheel's portability. Running `auditwheel show` on it shows that it relies on shared libraries:
 ```
-> % auditwheel show dist/pyoorb-1.2.0.post1+f37afd7.dirty-cp38-cp38-linux_x86_64.whl     

pyoorb-1.2.0.post1+f37afd7.dirty-cp38-cp38-linux_x86_64.whl is
consistent with the following platform tag: "linux_x86_64".

The wheel references external versioned symbols in these
system-provided shared libraries: libgcc_s.so.1 with versions
{'GCC_4.0.0', 'GCC_3.3', 'GCC_3.0', 'GCC_4.3.0', 'GCC_4.2.0'},
libc.so.6 with versions {'GLIBC_2.15', 'GLIBC_2.10', 'GLIBC_2.6',
'GLIBC_2.2.5', 'GLIBC_2.3.2', 'GLIBC_2.25', 'GLIBC_2.3.4',
'GLIBC_2.3', 'GLIBC_2.7', 'GLIBC_2.14', 'GLIBC_2.4', 'GLIBC_2.17'},
libgfortran.so.5 with versions {'GFORTRAN_8'}, libm.so.6 with versions
{'GLIBC_2.23', 'GLIBC_2.2.5', 'GLIBC_2.29', 'GLIBC_2.27'},
libquadmath.so.0 with versions {'QUADMATH_1.0'}

This constrains the platform tag to "linux_x86_64". In order to
achieve a more compatible tag, you would need to recompile a new wheel
from source on a system with earlier versions of these libraries, such
as a recent manylinux image.
```
In other words, users would need to install libgcc, libc, libgfortran, libm, and libquadmath to get this thing to work. Those aren't a _terrible_ list, they're pretty easy, to find, but it's still work. `auditwheel repair` may be able to make it actually self-contained, if I could get building to work on a manylinux system (I haven't tried).

### CODE REVIEW

File: python/setup.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+        src = "../lib/" + self.get_ext_filename(ext.name)
+        dst = self.get_ext_fullpath(ext.name)
+        self.mkpath(os.path.dirname(dst))
+        self.move_file(src, dst)
+
+
+with open("../VERSION", "r") as f:
+    version = f.read().strip()
+
```
Comment (spenczar) [inline python/setup.py:34]: This is... not my favorite thing to do. But at least it keeps the VERSION file authoritative.

File: python/setup.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
+import os
+import subprocess
+import pathlib
+from setuptools import setup, Extension
+from setuptools.command.build_ext import build_ext
+
+
+extension = Extension(
+    name="pyoorb",
+    sources=["pyoorb.f90", "pyoorb.pyf"],
```
Comment (spenczar) [inline python/setup.py:10]: NB these don't have any effect at all, since we just call `Make`. But setting non-empty `sources` is required by the Extension constructor, so I do it anyway.

**Comment by Siltala** [comment]
Nice to see work on this - thank you! I haven't taken a detailed look as of yet, but the new setup.py seems simple enough to understand.
I assume that further work will come as part of this PR given your todo list? I was also a bit concerned regarding portability - I don't think we can really get rid of the dependencies you state. It's not really a problem when making packages for more general package managers, but I don't know what's the best way to handle this. The `auditwheel repair` command you mention does seem promising in that sense.

**Comment by spenczar** [comment]
Happy to help!

> I assume that further work will come as part of this PR given your todo list?

Yep, I'd like to do those items. I can do them in this PR.

> I was also a bit concerned regarding portability - I don't think we can really get rid of the dependencies you state. It's not really a problem when making packages for more general package managers, but I don't know what's the best way to handle this. The `auditwheel repair` command you mention does seem promising in that sense.

Yeah, we'll see. I will try `auditwheel repair`, and if it works easily, great! If not, I propose that we add instructions that list the libraries users need to install if they are installing with pip; users who want a very simple installation method can use conda to get all those libraries automatically.

**Comment by mkelley** [comment]
Hi all, is this approach going to work?  It would be great for many users, but also for sbpy continuous integration testing (https://github.com/NASA-Planetary-Science/sbpy) if not elsewhere.  How would the data (planetary ephemerides, etc) for a pypi install be distributed for a pypi install?

**Comment by spenczar** [comment]
@mkelley I'm not sure whether this will work, and it might take someone with a better understanding of `f2py` than I to figure it out. I have not been able to make a truly portable Python Wheel which bundles the compiled code _and_ is able to find libraries like libgfortran. I don't really understand why, and may be doing something deeply wrong, but I have not yet succeeded in making a redistributable, standalone oorb for Python. I haven't given this much attention in the past month, and don't foresee returning to it too soon, but agree it really would be useful.

**Comment by mkelley** [comment]
@spenczar  Thanks for giving it a try, though.  It is beyond my expertise.

For what it is worth, I made a pip installable version that compiles the source (but IIRC, only with gfortran).  It is not as efficient as distributing binaries, but it works, at least for when I need it.  Maybe you can still salvage what you've done and use that approach?

https://github.com/mkelley/pyoorb-experiment

**Comment by Siltala** [comment]
Hi all,

Just posting to document that we do keep an eye on this PR. :)
This is also beyond my current expertise, and I have plenty of matters keeping me busy as it is, so I can't promise to work on this myself anytime soon. But I would be happy if someone could figure this out properly!

**Comment by spenczar** [comment]
I'm back onto this. Using [cibuildwheel](https://cibuildwheel.readthedocs.io/en/stable/), this is looking much more tractable, and I have portable wheels (I think?) working on my laptop currently. I'm more optimistic that they'll work, although I have to do some weird stuff like moving setup.py and a pyproject.toml to the top directory. 

One thing still to figure out is how to bundle the `de430.dat` ephemeris file into the wheel, but I think that should be relatively straightforward.

**Comment by spenczar** [comment]
Yeah, I can confirm that the Linux AMD64 wheels, at least, are working. That is, I can start a totally barebones Docker container:
```sh
$ docker run --name oorb_wheel_test --rm --platform linux/aarch64 -it ubuntu:focal /bin/bash
# Separate terminal, copy in the oorb `data` directory:
$ docker cp ~/src/oorb/data/ oorb_wheel_test:/tmp/oorb_data
```
In the container, I just install Python and then `pip install` the wheel - note no Fortran or lapack required:
``` sh
$ apt install python3.8-pip
$ pip install /tmp/pyoorb-1.0.2.post1+26f5352.dirty-cp38-cp38-manylinux_2_17_aarch64.manylinux2014_aarch64.whl
$ export OORB_DATA=/tmp/oorb_data
$ python3.8
```
And it works, I can propagate an orbit:
```python
>>> import pyoorb
>>> import numpy as np
>>> pyoorb.pyoorb.oorb_init()
>>> orbits = np.array([[0, 1.46905, 0.33435, np.deg2rad(14.3024),np.deg2rad(224.513), np.deg2rad(27.5419),np.deg2rad(324.697), 3, 51544.5, 1, 12.5, 0.15]],dtype=np.double, order='F')
>>> epoch = np.array([51232.23, 3], dtype=np.double, order='F')
>>> orb, err = pyoorb.pyoorb.oorb_propagation(in_orbits=orbits,in_epoch=epoch,in_dynmodel='N')
>>> err
0
>>> orb
array([[0.00000000e+00, 1.46902259e+00, 3.34239111e-01, 1.43017270e+01,
        2.24513002e+02, 2.75625816e+01, 1.51798113e+02, 3.00000000e+00,
        5.12322300e+04, 3.00000000e+00, 1.25000000e+01, 1.50000000e-01]])
``` 

I haven't yet done this testing routine for every version of Python and every architecture, but I think it should work with each of them.
---
This manual installation of the oorb data set is another thing to work on. I'd like to do that in a separate PR, though. I see two options for that:
- Include de430.dat and friends in the Python wheel. This makes the wheels ~100MB each, which is pretty big. On the other hand, oorb isn't usable without this, so maybe that's unavoidable.
 - Make a separate Python wheel that contains data files, so users do something like `pip install oorb-data`. I think this would require a few little changes to pyoorb to get it to look for that package and use it if it exists, but nothing too major. This means it takes two steps to install data, but data could be updated separately from the pyoorb code.

File: pyproject.toml [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+license = "GPL-3.0-only"
+
+[build-system]
+requires = ["setuptools", "wheel", "numpy"]
+
+[tool.cibuildwheel]
+before-build = """
+pip install numpy &&
+./configure gfortran opt --with-pyoorb --with-f2py=f2py --with-python=python
+"""
```
Comment (spenczar) [inline pyproject.toml:11]: Just to explain what's going on here: cibuildwheel lets you do pre-build work in two steps.

`before-all` is run just once for a target platform, across all Python versions. So this is used for installing system libraries.

`before-build` is run separately for each targeted Python version. So this is used when anything is Python version-dependent. That's true for us because f2py is linking to Python headers for a specific version.

So this config is used by every build. The ones below are platform-specific, eg using `yum` on the default linux builder, `brew` for the Mac one, and so on.

**Comment by spenczar** [comment]
@Siltala, I think this is ready for review. I cleaned up the git history. I added automatic tests that the wheels work, and that running the `test.py` file succeeds, for all targeted operating systems and architectures. The most recent run is over on my fork: https://github.com/spenczar/oorb/actions/runs/3906705980

File: .github/workflows/python_wheels.yml [hunk 4]
```diff
@@ -0,0 +1,195 @@
+name: Build Python Wheels
+
+# TODO: Shrink this down to pushes to master, perhaps. During development, this
+# needs to run more often to work out bugs.
+on:
+  - push
```
Comment (spenczar) [inline .github/workflows/python_wheels.yml:6]: @Siltala, let me know what you would like to see here. Building and testing the wheels is pretty slow, since aarch64 architectures need to be emulated. The whole thing takes about 2-3 hours.

The current workflow doesn't actually upload the built wheels anywhere. https://cibuildwheel.readthedocs.io/en/stable/deliver-to-pypi has some guides on how to do that. I think it might be something like this:
```yaml
  build_sdist:
    # Make a plain source distribution with nothing precompiled.
    runs-on: ubuntu-latest
    steps:
    - uses: actions/checkout@v3
    - name: Build SDist
      run: pipx run build --sdist
    - uses: actions/upload-artifact@v3
      with:
        name: sdist
        path: dist/*.tar.gz

  upload_all:
    # Upload all wheels and sdist to pypi
    needs: 
      - build_sdist
      - test_x86_wheels
      - test_linux_aarch64_wheels
    runs-on: ubuntu-latest
    if: github.event_name == 'release' && github.event.action == 'published'
    steps:
    - uses: actions/download-artifact@v3
      with:
        name: sdist
        path: ./sdist/
    - uses: actions/download-artifact@v3
      with:
        name: x86_wheels
        path: ./x86_wheels/
    - uses: actions/download-artifact@v3
      with:
        name: aarch64_wheels
        path: ./aarch64_wheels/
    - name: Combine artifacts
      run: |
        mkdir dist
        mv sdist/* dist
        mv x86_wheels/* dist
        mv aarch64_wheels/* dist
    - uses: pypa/gh-action-pypi-publish@v1.5.0
      with:
        user: __token__
        password: ${{ secrets.pypi_password }}
```

But of course I haven't tested that, because it needs a `pypi_password` to be set for the repository.

**Comment by mkelley** [comment]
This is great to see!  The package size might be larger than pypi allows.  Could it be packaged without the ephemeris file, and the file downloaded at install or runtime?

**Comment by spenczar** [comment]
> This is great to see!  The package size might be larger than pypi allows.  Could it be packaged without the ephemeris file, and the file downloaded at install or runtime?

Right, I am packaging it without the ephemeris currently. But making it all pip-installable would be nice. That could be done with an optional separate package. It could be an optional dependency so users might do 'pip install pyoorb[ephemdata]' to get it, for example.


The limit is 60MB after compression. It appears that all one has to do to get past the limit is make an issue on https://github.com/pypa/packaging-problems and the limit can be raised, and it looks like there isn't much fuss if you have a good reason, which I think we do.

**Comment by mkelley** [comment]
Got it, thanks again!

**Comment by rhiannonlynne** [comment]
I typically install from conda-forge instead, but it's nice to see things packaged up to sit in multiple types of repositories. And also definitely it's good to see that this should build on apple silicon /  os-arm64.

File: .github/workflows/python_wheels.yml [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+        name: Install Python 3.8
+        uses: actions/setup-python@v2
+        with:
+          python-version: 3.8
+
+      - if: ${{ matrix.python_version_tag == 'cp39-cp39' }}
+        name: Install Python 3.9
+        uses: actions/setup-python@v2
+        with:
+          python-version: 3.9
```
Comment (Siltala) [inline .github/workflows/python_wheels.yml]: I'm a little hesitant about these hardcoded version numbers since it sounds like something that needs manual maintenance as new Python versions come out. Is it possible to just parse the version number from `matrix.python_version_tag` and use that to install the relevant python setup? I think that would be less hassle free to keep running down the line.

File: setup.py [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+setup(
+    name="pyoorb",
+    version=deduce_version(),
+    ext_modules=[extension],
+    install_requires=["numpy"],
+    license="GPL3",
+    cmdclass={
+        "build_ext": PyoorbBuild,
+    }
```
[the remark this answers is not in the document]
Comment (Siltala) [reply setup.py:51]: I think that it would be best to do fill it in now rather than leave it for later.

Adapting the numpy one seems reasonable, we could do something among these lines:
I don't know about which email to use though. Perhaps the mailing list ( oorb@googlegroups.com )?
```
name='pyoorb',
maintainer="oorb developers",
maintainer_email="???",
description="An open-source orbit-computation package for Solar System objects. ",
long_description=Path("README.md").read_text(encoding="utf-8"),
long_description_content_type="text/markdown",
url="https://github.com/oorb/oorb",
author="Mikael Granvik et al.",
download_url="https://pypi.python.org/pypi/oorb",
project_urls={
"Bug Tracker": "https://github.com/oorb/oorb/issues",
"Source Code": "https://github.com/oorb/oorb",
 ```

File: .github/workflows/python_wheels.yml (hunk 5 above)
Comment (Siltala) [reply .github/workflows/python_wheels.yml]: Actually I see that the following commit changes this already

**Comment by Siltala** [comment]
So while I have no experience with actual python wheel builds or packaging, I think this looks mostly fine.
The only real concerns I have are the hardcoded python version numbers in the scripts because of (presumed) minor maintenance burden as new Python releases come out.
Also I think the commits could (and should) be cleaned up a little more. It seems strange for one commit to introduce new code only to have it removed in the following commits in a single PR i.e. commits "[Use cibuildwheel for building wheels](https://github.com/oorb/oorb/pull/136/commits/5c7c94ee1dd3729e491a1f0eb26e62037edebb94)" and [Add Github workflow to build wheels](https://github.com/oorb/oorb/pull/136/commits/64b4ab1ea6e3f0d78cd09a2eec74d7ccbe83ee3b). Maybe it would be simplest to just squash those commits together?

File: .github/workflows/python_wheels.yml (hunk 4 above)
Comment (spenczar) [reply .github/workflows/python_wheels.yml:6]: I'll be happy to help with the PyPI side of this. It ought to be simple, I hope; the hard part is building the wheels.

**Comment by spenczar** [comment]
@Siltala thanks for the review! I have cleaned up the git history a bit; let me know if you want me to squash it all the way down to a single commit, though. 

I think the Python versions are reduced as far as is possible with Github Actions. It's unfortunate that they'd need to be updated in multiple places, but GH Actions don't provide tools for abstracting across multiple jobs.

---

For Apple M1 users, there's no good option presented in this PR. They'd have to build from source. As of this PR, that would be done by cloning the repo and building manually.

An alternative in the Python world is a "source distribution." When a user does `pip install pyoorb`, pip will first check PyPI to see if there is a pre-built binary available for their Python version and platform. If not, it will check PyPI for a "sdist" or "source distribution," and download that instead, and attempt to build it using the local toolchain.

So we could make an sdist, and push it up to PyPI, which would bundle in all the fortran code. Users on Apple M1 would need to have gfortran installed already, but then they could `pip install pyoorb` and it ought to work.

I'd like to do that in a separate PR, since it's pretty different from what I have here. I have preliminary work over in https://github.com/spenczar/oorb/compare/build_python_wheels...spenczar:oorb:python_sdist.

File: setup.py [hunk 7]
```diff
@@ -41,7 +42,19 @@ def deduce_version():
...
-    name="pyoorb",
+    name='pyoorb',
+    maintainer="oorb developers",
+    maintainer_email="oorb@googlegroups.com",
+    description="An open-source orbit-computation package for Solar System objects. ",
+    long_description=Path("README.md").read_text(encoding="utf-8"),
+    long_description_content_type="text/markdown",
+    url="https://github.com/oorb/oorb",
+    author="Mikael Granvik et al.",
+    download_url="https://pypi.python.org/pypi/oorb",
```
Comment (Siltala) [inline setup.py:53]: Actually, here's a small nitpick - maybe we want to call it pyoorb in pypi which would make https://pypi.python.org/pypi/pyoorb the proper `download_url` here. My fault for suggesting this one earlier!

**Comment by Siltala** [comment]
Yes, all that sounds good. I think it's fine to leave the sdist part for a later PR, and I think the commit history looks acceptable now.

File: setup.py (hunk 7 above)
Comment (spenczar) [reply setup.py]: Oops, good catch. Fixed.

**Comment by spenczar** [comment]
@Siltala I think this is ready, do you have any further comments?
